src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -113,6 +115,245 @@ void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+    return UCS_OK;
+#else
+    ctx->support = false;
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_list_free(struct uct_rdmacm_cm_reserved_qpn_context *ctx)
+{
+    ucs_status_t status = UCS_OK;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk, *tmp;
+
+    pthread_mutex_lock(&ctx->mutex);
+
+    ucs_list_for_each_safe(blk, tmp, &ctx->blk_list, entry) {
+        ucs_list_del(&blk->entry);
+        status = uct_rdmacm_cm_reserved_qpn_blk_free(blk);
+        if (status != UCS_OK) {
+            goto end;
+        }
+    }
+
+end:
+    pthread_mutex_unlock(&ctx->mutex);
+#endif
+
+    return status;
+}
+
+static void uct_rdmacm_cm_reserved_qpn_context_cleanup(uct_rdmacm_cm_t *cm)
+{
+    struct uct_rdmacm_cm_reserved_qpn_context *ctx;
+    ucs_status_t status;
+
+    kh_foreach_value(&cm->ctxs, ctx, {
+        status = uct_rdmacm_cm_reserved_qpn_blk_list_free(ctx);
+        if (status != UCS_OK) {
+            ucs_error("qpn_context_cleanup failed: %m");
+        }
+        ucs_free(ctx);
+    });
+
+    kh_destroy_inplace(uct_rdmacm_cm_qpn_contexts, &cm->ctxs);
+}
+
+ucs_status_t uct_rdmacm_cm_get_reserved_qpn_context(uct_rdmacm_cm_t *cm,
+                                                    struct ibv_context *verbs,
+                                                    struct uct_rdmacm_cm_reserved_qpn_context **ctx_p)
+{
+    struct uct_rdmacm_cm_reserved_qpn_context *ctx;
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(uct_rdmacm_cm_qpn_contexts, &cm->ctxs,
+                  ibv_get_device_guid(verbs->device), &ret);
+    if (ret == -1) {
+        ucs_error("cm %p: cannot allocate hash entry for reserved qpn context", cm);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (ret == 0) {
+        /* already exists so use it */
+        ctx = kh_value(&cm->ctxs, iter);
+    } else {
+        /* Create a reserved qpn context */
+        ctx = ucs_malloc(sizeof(struct uct_rdmacm_cm_reserved_qpn_context),
+                         "reserved_qpn_context");
+        if (ctx == NULL) {
+            kh_del(uct_rdmacm_cm_qpn_contexts, &cm->ctxs, iter);
+            ucs_error("cm %p: malloc reserved qpn context failed: %m", cm);
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        pthread_mutex_init(&ctx->mutex, NULL);
+        ucs_list_head_init(&ctx->blk_list);
+        uct_rdmacm_cm_query_reserved_qpn_cap(verbs, ctx);
+        uct_rdmacm_cm_query_reserved_qpn_cap_2(verbs, ctx);
+
+        kh_value(&cm->ctxs, iter) = ctx;
+    }
+
+    *ctx_p = ctx;
+    return UCS_OK;
+}
+
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_t *cm,
+                                     struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_reserved_qpn_context *ctx,
+                                     struct uct_rdmacm_cm_reserved_qpn_blk **blk_p)
+{
+    ucs_status_t status = UCS_OK;
```

Comment:
so, if **not** `HAVE_DECL_MLX5DV_IS_SUPPORTED` the func returns OK, is it expected? I guess this function should be compiled out at all then
the same for `uct_rdmacm_cm_reserved_qpn_blk_free`

Reply:
@evgeny-leksikov Actually if not HAVE_DECL_MLX5DV_IS_SUPPORTED the func won't be called at all because we checked reserved_qpn_context->support before calling it. But seems return UNSUPPORTED is better than return OK. I changed it to return UNSUPPORTED.
We can compile out this function and also all Reserved QPN related functions, but then we need add more HAVE_DECL_MLX5DV_IS_SUPPORTED checking at the places calling these functions and defining the related structs. Currently **at the high level(create/destroy QPN)** we unify to use reserved_qpn_context->support to judge dummy QP or Reserved QPN(run time checking), we don't want to use both reserved_qpn_context->support and HAVE_DECL_MLX5DV_IS_SUPPORTED(run time and compile time checking).